### PR TITLE
Add support for function-based velocity advection on constant and higher order discontinuous variables

### LINKED ISSUE #183: Fix function-based InflowBC and Outflow postprocessor in PrecursorAction
## Issue
Moltres supports precursor drift with uniform, function-based, or Navier-Stokes coupled velocities.

The current implementation of outlet postprocessors and inlet boundary conditions in `PrecursorAction.C` with **function-based velocities** and **precursor looping**:
1. does not impose an outlet postprocessor for calculating the precursor concentration at the outlet; and
2. imposes the wrong type of inlet boundary condition without any velocity values provided.

These errors lead to:
1. MooseError stating that ` Reporter name "Outlet_Average_pre1_/value" with type "double" is not declared.`; and
2. homogeneous Dirichlet boundary conditions being imposed on the inlet.

## How to reproduce

**Issue 1:**
Replace the lines 26-28 in `$MOLTRES/tests/pre/sub.i`:
``` 
    u_def = ${flow_velocity}
    v_def = 0
    w_def = 0
```
with:
```
    constant_velocity_values = false
    u_func = vel_func
    v_func = 0
    w_func = 0
```
Add the following code block anywhere outside other input sections:
```
[Functions]
  [./vel_func]
    type = ParsedFunction
    value = ${flow_velocity}
  []
[]
```
The velocity function imposes the same constant velocity value as the original input file, but the `pre_loop.i` test fails due to the erroneous function-based velocity boundary conditions imposed by `PrecursorAction.C`.

**Issue 2:**
In addition to the code changes to reproduce Issue 1, replace line 401 in `PrecursorAction.C`:
```
  if (getParam<bool>("constant_velocity_values"))
```
with:
```
  if (getParam<bool>("constant_velocity_values") || !isParamValid("uvel"))
```
This step fixes Issue 1 but introduces Issue 2.

## How to fix
1. Fix the outlet postprocessor implementation logic in `PrecursorAction.C` as suggested above.
2. Create an inflow boundary condition, similar to `PostprocessorInflowBC`, for function-based velocities and fix the inlet boundary condition implementation logic in `PrecursorAction.C`.
3. Create a test for precursor looping with function-based velocities.

### LINKED ISSUE #265: Extend PrecursorAction to support modeling with higher order precursor variables
The PrecursorAction action object for setting up delayed neutron precursor modeling parameters currently only works for CONSTANT order precursor variables. Higher order variables would improve the solution resolution on the same mesh. However, modeling with higher order variables requires additional advection kernels in addition to the existing advection dgkernels.

### PR DESCRIPTION
**Pull request by smpark7**
## Description
`PostprocessorVelocityFunctionInflowBC` and `VelocityFunctionConservativeAdvection` are similar to `PostprocessorInflowBC` and `ConservativeAdvection`, but with function-based velocity profiles.

`PostprocessorPenaltyDirichletBC` is similar to `PenaltyDirichletBC`, but with a postprocessor to set the boundary value.

`PostprocessorVelocityFunctionInflowBC` is required to fix arfc/moltres#183.

`VelocityFunctionConservativeAdvection` and `PostprocessorPenaltyDirichletBC` are required to fix arfc/moltres#265.

## Changes
- Add `PostprocessorVelocityFunctionInflowBC`
- Add `VelocityFunctionConservativeAdvection`
- Add `PostprocessorPenaltyDirichletBC`

### CODE REVIEW

File: doc/content/source/bcs/PostprocessorVelocityFunctionInflowBC.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# PostprocessorVelocityFunctionInflowBC
+
+!alert construction title=Undocumented Class
```
Comment (LukeSeifert) [inline doc/content/source/bcs/PostprocessorVelocityFunctionInflowBC.md]: Is documentation going to be added in a future PR?

File: doc/content/source/kernels/VelocityFunctionConservativeAdvection.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# VelocityFunctionConservativeAdvection
+
+!alert construction title=Undocumented Class
```
Comment (LukeSeifert) [inline doc/content/source/kernels/VelocityFunctionConservativeAdvection.md]: Same here

File: src/bcs/PostprocessorPenaltyDirichletBC.C [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  : IntegratedBC(parameters),
+    _p(getParam<Real>("penalty")),
+    _pp(getPostprocessorValue("postprocessor"))
+{
+}
+
+Real
+PostprocessorPenaltyDirichletBC::computeQpResidual()
+{
+  return _p * _test[_i][_qp] * (-_pp + _u[_qp]);
```
Comment (LukeSeifert) [inline src/bcs/PostprocessorPenaltyDirichletBC.C]: These variable names don't seem great to me, how come something longer isn't used?

**Review by LukeSeifert — COMMENTED** [review]
Looks good! I only had some questions on the documentation and the variable names. I have noticed when looking at MOOSE source code it uses really short variable names like `_t`, why is that?

**Comment by smpark7** [comment]
@LukeSeifert Thanks for the quick review! I added documentation to the markdown files and I renamed the variables to `_penalty` and `_postprocessor`. I think the MOOSE team used to be more lax about variables names. To be fair the names usually quite obvious (e.g. _t usually means time). As the framework grew over the years, naming conflicts have cropped up. However, it can be difficult to "rectify" some of the names without impacting a large number of users and MOOSE-based apps downstream.

Feel free to merge if you approve!

**Review by LukeSeifert — APPROVED** [review]
Looks good!

**Comment by cticenhour** [comment]
> I think the MOOSE team used to be more lax about variables names.

@LukeSeifert @smpark7 I have no doubt that we might have been a bit more lax in the past. 😄 The decision on `_t` was a decent bit before my time though. I _would like_ to think that short-named spatial and temporal variables that get inherited (like `_t`) come from how one naturally writes a simple function down on paper (i.e., `f(x,y,z,t) = 5xt`). I should ask one day. For my code, I try to make my object variable names longer so as to be more descriptive. There are almost certainly many places we could fix up in MOOSE and many apps though 😆
